### PR TITLE
feature: oss-680 upload graphql command

### PR DESCRIPTION
**Pull request by fireridlle**
### Notes
[OSS-Upload schema from the shell](https://faunadb.atlassian.net/browse/OSS-680)
[Issue "Upload schema from the shell"](https://github.com/fauna/fauna-shell/issues/67)

### How to test

USAGE
  $ fauna upload-graphql-shema GRAPHQLPATH

ARGUMENTS
  GRAPHQLPATH  Path to GraphQL schema

OPTIONS
  --graphqlHost=graphqlHost  [default: https://graphql.fauna.com] The Fauna GraphQL API host
  --mode=merge|override      [default: merge] Upload mode
  --secret=secret            FaunaDB secret key

EXAMPLES
  $ fauna upload-graphql-schema ./schema.gql
  $ fauna upload-graphql-schema ./schema.gql --mode override

### CODE REVIEW

**Review by parkhomenko — APPROVED** [review]

**Review by marrony — COMMENTED** [review]
How this integrates with `~/.fauna-shell` or `--endpoint` argument? I see this is getting the secret from there, but suppose I have a default endpoint the points to localhost, it will get the secret from there and try to upload the file on https://graphql.fauna.com

**Comment by fireridlle** [comment]
> How this integrates with `~/.fauna-shell` or `--endpoint` argument? I see this is getting the secret from there, but suppose I have a default endpoint the points to localhost, it will get the secret from there and try to upload the file on https://graphql.fauna.com

what do you mean by integration with `~/.fauna-shell`?

`endpoint` arguments is for db.
we have the same approach for vscode plugin

File: src/commands/upload-graphql-shema.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+  allowedExt = ['.graphql', '.gql']
+
+  async run() {
+    try {
+      const { graphqlPath } = this.args
+      const { mode, graphqlHost } = this.flags
+
+      if (!this.allowedExt.includes(path.extname(graphqlPath))) {
+        errorOut(
+          'You must provide a Graphql file (`.graphql` or `.gql`) to upload a schema'
```
Comment (faunaee) [inline src/commands/upload-graphql-shema.js:18]:
```suggestion
          'Your GraphQL schema file must include the `.graphql` or `.gql` extension.'
```

File: src/commands/upload-graphql-shema.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        )
+      }
+
+      const {
+        connectionOptions: { secret },
+      } = await this.getClient()
+
+      console.info(`UPLOADING SCHEMA (mode=${mode}): ${graphqlPath}`)
+      const text = await fetch(`${graphqlHost}/import?mode=${mode}`, {
```
Comment (faunaee) [inline src/commands/upload-graphql-shema.js:27]: Can we trim trailing whitespace and slashes from the URL provided for the `graphqlHost`?

File: src/commands/upload-graphql-shema.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+      console.info('RESPONSE:')
+      console.info(text)
+    } catch (error) {
+      errorOut(error)
+    }
+  }
+}
+
+UploadGraphQLSchemaCommand.description = 'Upload GraphQL Schema'
```
Comment (faunaee) [inline src/commands/upload-graphql-shema.js:41]:
```suggestion
UploadGraphQLSchemaCommand.description = 'Upload GraphQL schema'
```

**Review by faunaee — CHANGES_REQUESTED** [review]
Just a few minor changes.

I agree with Marrony though; there should be a way to store the `graphqlHost` in `~/.fauna-shell` so that only `--endpoint` needs to be specified.

**Comment by marrony** [comment]
@fireridlle if you do not specify `--secret` argument, `~/.fauna-shell` is the place it will search for it, based on the default `endpoint`.

File: src/commands/upload-graphql-shema.js (hunk 1 above)
Comment (fireridlle) [reply src/commands/upload-graphql-shema.js:18]: @faunaee i copied this text from vscode plugin https://github.com/fauna/vscode/blob/master/src/uploadGraphqlSchemaCommand.ts#L16
should it be consistent?

Comment (faunaee) [reply src/commands/upload-graphql-shema.js:18]: I would say so, yes.

**Comment by faunaee** [comment]
I've just tried the new command. I can't make it work with the Fauna Dev Docker image, because I can't specify the port for the GraphQL API. See https://docs.fauna.com/fauna/current/integrations/dev#ports

Can we add a `graphqlPort` option, both for command-line configuration, and as a config field in `~/.fauna-shell` ?

**Comment by fireridlle** [comment]
> I've just tried the new command. I can't make it work with the Fauna Dev Docker image, because I can't specify the port for the GraphQL API. See https://docs.fauna.com/fauna/current/integrations/dev#ports
> 
> Can we add a `graphqlPort` option, both for command-line configuration, and as a config field in `~/.fauna-shell` ?

done

**Comment by faunaee** [comment]
This works:

```
$ bin/run upload-graphql-schema --graphqlHost=localhost --graphqlPort=8084 ./schema.gql
UPLOADING SCHEMA (mode=merge): ./schema.gql
RESPONSE:
Schema imported successfully.
Use the following HTTP header to connect to the FaunaDB GraphQL API:
{ "Authorization": "Bearer fnADTHyV-ZACDBC9ROEi9kW8DDvL7V4ziOK5m3lH" }
```

With this entry in `~/.fauna-shell`:

```
[test]
domain=localhost
port=8443
scheme=http
secret=fnAEMRODcCACANIiSnji7-BEalUcbAufrmGrgN8m
graphqlHost=localhost
graphqlPort=8084
```

This does not work:

```
$ bin/run upload-graphql-schema --endpoint=test ./schema.gql
UPLOADING SCHEMA (mode=merge): ./schema.gql
FetchError: request to http://graphql.fauna.com:443/import?mode=merge failed, reason: read ECONNRESET
    at ClientRequest.<anonymous> (~/repos/fauna-shell/node_modules/node-fetch/lib/index.js:1461:11)
```

Notice that neither the `graphqlHost` nor `graphqlPort` settings have been applied, yet the `scheme` setting has.

**Comment by faunaee** [comment]
`replace` mode exists: https://docs.fauna.com/fauna/current/api/graphql/endpoints#modes

But this does not work:

```
$ bin/run upload-graphql-schema --graphqlHost=localhost --graphqlPort=8084 ./schema.gql --mode=replace
 ›   Error: Expected --mode=replace to be one of: merge, override
 ›   See more help with --help
```

File: README.md [hunk 4]
```diff
@@ -350,17 +350,24 @@ domain=127.0.0.1
 port=8443
 scheme=http
 secret=the_secret
+graphqlHost=127.0.0.1
+graphqlPort=443
+
```
Comment (faunaee) [inline README.md:355]: Do we need two blank lines here?

File: README.md [hunk 5]
```diff
@@ -350,17 +350,24 @@ domain=127.0.0.1
 port=8443
 scheme=http
 secret=the_secret
+graphqlHost=127.0.0.1
+graphqlPort=443
```
Comment (faunaee) [inline README.md:354]:
```suggestion
graphqlPort=8084
```
Just to match up with the recommended Fauna Dev configuration: https://docs.fauna.com/fauna/current/integrations/dev#ports

File: README.md [hunk 6]
```diff
@@ -350,17 +350,24 @@ domain=127.0.0.1
 port=8443
 scheme=http
 secret=the_secret
```
Comment (faunaee) [inline README.md:352]:
```suggestion
secret=secret
```
The default secret for Fauna Dev is `secret`. We don't make that a "secret": we show that in the last step of this procedure: https://docs.fauna.com/fauna/current/integrations/dev#run

**Review by faunaee — CHANGES_REQUESTED** [review]
Just a few more tweaks.

**Review by faunaee — APPROVED** [review]
LGTM! 👍

**Pull request merged**